### PR TITLE
Entra ID integration boilerplate

### PR DESCRIPTION
**Pull request by justinas**
Split off from https://github.com/gravitational/teleport/pull/40556 . Depends on https://github.com/gravitational/teleport/pull/40997

Introduces validation for types introduced in https://github.com/gravitational/teleport/pull/40997 , as well as modifications to OIDC/JWT logic required to sign tokens for Entra APIs. See https://github.com/gravitational/teleport/pull/40556 for more details.

### CODE REVIEW

File: api/types/integration.go [hunk 1]
```diff
@@ -290,7 +354,8 @@ func (ig *IntegrationV1) MarshalJSON() ([]byte, error) {
 	d := struct {
 		ResourceHeader `json:""`
 		Spec           struct {
-			AWSOIDC AWSOIDCIntegrationSpecV1 `json:"aws_oidc"`
+			AWSOIDC   AWSOIDCIntegrationSpecV1   `json:"aws_oidc"`
```
Comment (jakule) [inline api/types/integration.go:357]: You should probably add `omitempty` since now only one will be set

File: api/types/integration_test.go [hunk 2]
```diff
@@ -104,9 +119,7 @@ func TestIntegrationCheckAndSetDefaults(t *testing.T) {
 					nil,
 				)
 			},
-			expectedErrorIs: func(err error) bool {
-				return trace.IsBadParameter(err)
-			},
+			expectedErrorIs: trace.IsBadParameter,
```
Comment (jakule) [inline api/types/integration_test.go:122]: 👍

File: lib/integrations/azureoidc/token_generator.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	privateKey, err := services.GetJWTSigner(signer, ca.GetClusterName(), clock)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	token, err := privateKey.SignEntraOIDC(jwt.SignParams{
+		Audience: azureDefaultJWTAudience,
+		Subject:  "teleport-azure", // TODO(justinas): consider moving this to a constant or a field in the integration settings
```
Comment (jakule) [inline lib/integrations/azureoidc/token_generator.go:71]: Should this be addressed before merging?

File: lib/jwt/jwk_test.go [hunk 4]
```diff
@@ -34,3 +38,32 @@ func TestMarshalJWK(t *testing.T) {
 	// Required for integrating with AWS OpenID Connect Identity Provider.
 	require.Equal(t, "sig", jwk.Use)
 }
+
+func TestKeyID(t *testing.T) {
+	t.Run("deterministic", func(t *testing.T) {
```
Comment (jakule) [inline lib/jwt/jwk_test.go:43]: Those can be two separate tests

File: lib/jwt/jwk_test.go [hunk 5]
```diff
@@ -34,3 +38,32 @@ func TestMarshalJWK(t *testing.T) {
 	// Required for integrating with AWS OpenID Connect Identity Provider.
 	require.Equal(t, "sig", jwk.Use)
 }
+
+func TestKeyID(t *testing.T) {
```
Comment (jakule) [inline lib/jwt/jwk_test.go:42]:
```suggestion
func TestKeyID(t *testing.T) {
    t.Parallel()
```

**Review by jakule — APPROVED** [review]

File: lib/integrations/azureoidc/token_generator.go (hunk 3 above)
Comment (justinas) [reply lib/integrations/azureoidc/token_generator.go:71]: Shouldn't be a problem later: as long as we have no user-facing way to onboard this, there's nothing to break 🙂 Left it TODO as I'm not sure how to handle this, but considering 1 ID provider = 1 subject for purposes of Entra app auth, constant should be okay I guess.

File: lib/jwt/jwk_test.go [hunk 6]
```diff
@@ -34,3 +38,34 @@ func TestMarshalJWK(t *testing.T) {
 	// Required for integrating with AWS OpenID Connect Identity Provider.
 	require.Equal(t, "sig", jwk.Use)
 }
+
+func TestKeyIDHasConsistentOutputForAnInput(t *testing.T) {
```
Comment (strideynet) [inline lib/jwt/jwk_test.go:42]: It'd be nice to see a test which takes a known public key and asserts that it produces a fixed KeyID - that'd let us ensure that we don't accidentally introduce a regression that causes the key ID in an already issued JWT to mismatch the key ID in the jwks.

**Review by strideynet — APPROVED** [review]

File: lib/jwt/jwk_test.go (hunk 6 above)
Comment (justinas) [reply lib/jwt/jwk_test.go:43]: Good idea, added in c44c7c2d1441883c6ee8247669e39e136baad3a7.

File: api/types/integration.go [hunk 7]
```diff
@@ -128,14 +155,19 @@ func (s *IntegrationSpecV1) CheckAndSetDefaults() error {
 		if err != nil {
 			return trace.Wrap(err)
 		}
+	case *IntegrationSpecV1_AzureOIDC:
+		err := integrationSubKind.CheckAndSetDefaults()
```
Comment (tigrato) [inline api/types/integration.go:159]:
```suggestion
		err := integrationSubKind.Validate()
```

File: api/types/integration.go [hunk 8]
```diff
@@ -160,6 +192,21 @@ func (s *IntegrationSpecV1_AWSOIDC) CheckAndSetDefaults() error {
 	return nil
 }
 
+// CheckAndSetDefaults validates the configuration for Azure OIDC integration subkind.
+func (s *IntegrationSpecV1_AzureOIDC) CheckAndSetDefaults() error {
```
Comment (tigrato) [inline api/types/integration.go:196]:
```suggestion
// Validate validates the configuration for Azure OIDC integration subkind.
func (s *IntegrationSpecV1_AzureOIDC) Validate() error {
```

File: api/types/plugin.go [hunk 9]
```diff
@@ -609,6 +621,17 @@ func (c *PluginOAuth2AccessTokenCredentials) CheckAndSetDefaults() error {
 	return nil
 }
 
+func (c *PluginEntraIDSettings) CheckAndSetDefaults() error {
```
Comment (tigrato) [inline api/types/plugin.go:624]:
```suggestion
func (c *PluginEntraIDSettings) Validate() error {
```

**Review by tigrato — APPROVED** [review]

**Pull request merged**